File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/proposer/common/AbstractFairnessProposer.java
package cz.fb.manaus.reactor.betting.proposer.common;

import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.reactor.betting.BetContext;
import cz.fb.manaus.reactor.betting.proposer.PriceProposer;
import cz.fb.manaus.reactor.betting.validator.ValidationResult;
import cz.fb.manaus.reactor.price.PriceService;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.OptionalDouble;
import java.util.function.Function;

public abstract class AbstractFairnessProposer implements PriceProposer {

    private final Side side;
    private final Function<BetContext, Double> downgradeStrategy;
    @Autowired
    private PriceService priceService;

    public AbstractFairnessProposer(Side side, Function<BetContext, Double> downgradeStrategy) {
        this.side = side;
        this.downgradeStrategy = downgradeStrategy;
    }

    @Override
    public ValidationResult validate(BetContext context) {
        return ValidationResult.of(context.getFairness().get(side).isPresent());
    }

    @Override
    public OptionalDouble getProposedPrice(BetContext context) {
        var fairness = context.getFairness().get(side);
        var bestPrice = context.getRunnerPrices().getHomogeneous(side).getBestPrice().get();
        var fairPrice = priceService.getFairnessFairPrice(bestPrice.getPrice(), fairness.getAsDouble());
        var downgradeFraction = downgradeStrategy.apply(context);
        return OptionalDouble.of(priceService.downgrade(fairPrice, downgradeFraction, context.getSide()));
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/proposer/common/AbstractFairnessProposer.kt
package cz.fb.manaus.reactor.betting.proposer.common

import cz.fb.manaus.core.model.Side
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.proposer.DowngradeStrategy
import cz.fb.manaus.reactor.betting.proposer.PriceProposer
import cz.fb.manaus.reactor.betting.validator.ValidationResult
import cz.fb.manaus.reactor.price.PriceService
import org.springframework.beans.factory.annotation.Autowired

abstract class AbstractFairnessProposer(private val side: Side, private vararg val downgradeStrategies: DowngradeStrategy) : PriceProposer {
    @Autowired
    private lateinit var priceService: PriceService

    override fun validate(event: BetEvent): ValidationResult {
        return if (event.metrics.fairness[side] != null) ValidationResult.ACCEPT else ValidationResult.REJECT
    }

    override fun getProposedPrice(event: BetEvent): Double {
        val fairness = event.metrics.fairness[side]!!
        val bestPrice = event.runnerPrices.getHomogeneous(side).bestPrice!!
        val fairPrice = priceService.getFairnessFairPrice(bestPrice.price, fairness)
        val provider = event.account.provider
        val strategy = downgradeStrategies.find(provider::matches) ?: error("no such strategy")
        val downgradeFraction = strategy(event)
        return priceService.downgrade(fairPrice, downgradeFraction, event.side)
    }
}
